byway/src/main/java/com/google/sps/data/Trip.java

```diff
@@ -0,0 +1,89 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.Entity;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+
+/**
+ * A class to make a Trip type, containing
+ * a specific id, destinations, interests and routes 
+ * to be used when updating a trip.
+ */
+public final class Trip {
+
+  public static final String DATASTORE_ENTITY_KIND = "Trip";
+
+  private final String id;
+  private final String start;
+  private final ArrayList<String> destinations;
+  private final ArrayList<String> interests;
+  private final ArrayList<String> route;
+
+  public Trip(String id, String start, Collection<String> destinations,
+              Collection<String> interests, Collection<String> route) {
+    this.id = checkNotNull(id, "id");
+    this.start = checkNotNull(start, "start");
+    Collection<String> validDestinations = checkNotNull(destinations, "destinations");
+    Collection<String> validInterests = checkNotNull(interests, "interests");
+    Collection<String> validRoute = checkNotNull(route, "route");
+    this.destinations = new ArrayList<String>(validDestinations);
+    this.interests = new ArrayList<String>(validInterests);
+    this.route = new ArrayList<String>(validRoute);
+  }
+
+  public ArrayList<String> getInterests() {
+    return (ArrayList<String>) Collections.unmodifiableList(this.interests);
+  }
+
+  public ArrayList<String> getRoute() {
+    return (ArrayList<String>) Collections.unmodifiableList(this.route);
+  }
+
+  public ArrayList<String> getDestinations() {
+    return (ArrayList<String>) Collections.unmodifiableList(this.destinations);
+  }
+
+  public String getTripId() {
+    return this.id;
+  }
+
+  public String getStart() {
+    return this.start;
+  }
+
+  public static Trip FromEntity(Entity tripEntity) {
+    checkArgument(tripEntity.getKind().equals(DATASTORE_ENTITY_KIND),
+      "Wrong Entity kind. Expected %s, received %s", DATASTORE_ENTITY_KIND, tripEntity.getKind());
+    String id = checkNotNull((String) tripEntity.getProperty("id"),
```

Comment:
If you want something to be the canonical "ID" of an entity, it would be good to make it the Key of the entity, that way you can be sure you can't accidentally have to multiple entities with the same ID.

Note that for serialization to JSON, you can still store the "id" as a string, but you might find it useful to have a
```java
Key getKey() {
   return new Key(DATASTORE_ENTITY_KIND, id);
}
```

Reply:
We decided to rename this id variable throughout the app to reference the trip key itself rather than labelling it as id, since we could mistakenly be referring to the key's id. I implemented this function as well but with a `stringToKey` method instead, effectively doing the same thing.

Updated

Reply:
There is a specific field on the Entity for its key:

https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/Entity.html#getKey--

If you use the actual datastore key of the entity, you are guaranteed that no other entity has the same key as this entity. If you store a string in your own property, you could have multiple entities with the same value for `tripEntity.getProperty("id")`.

Using `id = KeyFactory.keyToString(tripEntity.getKey())` is fine with me if you want to have the ID just be the encoded key value and that's probably easier.

However, what I was suggesting above was that you can also use something like `id = tripEntity.getKey().getName()`, assuming when you create the entity you get the key using `createKey(DATASTORE_ENTITY_KIND, theIdYouWantForThisEntity)`. Then whenever you get a request with a tripId, you turn the tripId back into a datastore key the same way, using `createKey(DATASTORE_ENTITY_KIND, tripId)`.

This approach, using `createKey(DATASTORE_ENTITY_KIND, id)` is especially useful for the `User` type, since you would be getting a User from AppEngine, and can just use `createKey(DATASTORE_ENTITY_KIND, user.getUserId())` as the key for the entity, re-using the [ID provided by AppEngine](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/users/User.html#getuserid) as the ID in datastore.

Reply:
Both ways are implementable/relatively straight forward. I first set it up as passing the key's id, converting to long, and creating a key, `createKey(DATASTORE_ENTITY_KIND`, id)`, and it worked fine. But it felt like an extra step, converting strings to longs and checking for number format exceptions, then creating the key. With the other method, it only converts two-way, string to key, and the parameters are passed in as strings so it avoids getting the correct type.

This approach applies to both the Trip and User class. I think `createKey()` required the id of the key itself rather than `user.getUserId()`. I'm not sure of the distinction between both, since the key's id references the user, but so should userId which is based on the same user. Also, wouldn't the appengine `User` class interfere with the custom `User` class being made here? For now, I'll implement it with the keyToString method rather than use id, and can update it with ids if problems arise

Reply:
When you create a new entity, it will always have *a* key, and you have 3 choices for how to decide on a key for it.

1) Create an entity with a Kind only. When you insert it into the datastore, it will get a numeric (`long`) ID value chosen automatically. To do this, you construct the entity with `new Entity(DATASTORE_ENTITY_KIND)` and then when you call `datastore.put(entity)` the key will be assigned, and can be retrieved with `entity.getKey()`.
2) Choose your own value for the `long` ID. You create your Key ahead of time via the `KeyFactory`, and create the entity with it. Example: `new Entity(KeyFactory.createKey(DATASTORE_ENTITY_KIND, yourChosenId))`. When you use `datastore.put(entity)`, the ID will be left with the value you chose.
3) Choose your own value for the `String` `name` of the key, and create the entity with it. Example: `new Entity(KeyFactory.createKey(DATASTORE_ENTITY_KIND, yourChosenName))`. When you use `datastore.put(entity)`, the `name` will be used, and no numeric `id` will be assigned.

Note that when you are using option (2) or (3), you have to make sure that the number or string that you choose really is unique, otherwise you will replace an existing entity with that key when you use `datastore.put(entity)`, rather than add a new entity.

Option (3) is useful for working with the AppEngine `UserService`, because the `UserService` provides a unique ID per-user already, so it's pretty straightforward to just use that as your the Key's `name` rather than having the datastore choose a numeric `id` for you. You'll find that this is especially convenient when it comes to retrieving the user entity for a particular user. Example below, but first, let me address this:

> Also, wouldn't the appengine User class interfere with the custom User class being made here?

No, but also yes, sort of.

In Java, your classes have both a name and a package. The full name of a Java class is really the combination of the package name and the class name. Since your `User` class is declared in a file with `package com.google.sps.data`, the full name of the `User` class is actually `com.google.sps.data.User`. Meanwhile, the `User` class from AppEngine is actually `com.google.appengine.api.users.User`. When you write `import com.google.appengine.api.users.User`, what you're saying is "when I write `User`, what I mean is `com.google.appengine.api.users.User`". Imports are technically optional! You could *technically* write all of your code using only the full names of classes including the package. Imports exist primarily to make code easier to read and write for humans. Also, if you are writing a class in a particular package, any other classes in that package are available by default as if you imported them.

So what this means is that *technically*, the `com.google.appengine.api.users.User` will not interfere with your `User` class, because you can write `User` when you mean `com.google.sps.data.User`, and then spell-out the full `com.google.appengine.api.users.User`. But practically speaking, yeah this kind of interferes in your code, since now you have to write a really long name when using one of them. So practically speaking, it might be a good idea to rename your `User` class to something else, like `UserInfo`, so you can have both names imported at the same time.

----

Anyway, back to how you would use the user ID as a key.

You can imagine writing code that does something sort of like this in order to get or create the User entity:

```java
com.google.appengine.api.users.User user = userService.getCurrentUser();
if (user != null) {
  Key userKey = KeyFactory.createKey(DATASTORE_ENTITY_KIND, user.getUserId());
  Entity userEntity;
  try {
    userEntity = datastore.get(userKey);
  } catch (EntityNotFoundException ex) {
    userEntity = new Entity(userKey);
    userEntity.setProperty("email", user.getEmail());
    userEntity.setProperty("tripIds", new ArrayList<Key>()); // note ArrayList<Key>; when referring to datastore entites from other datastore entities, I'd recommend using explicit Key rather than a string formatted value.
    datastore.put(userEntity);
  }
  return fromEntity(userEntity);
}
return null;
```

Reply:
I see, so the userId would make sense since we can input it directly for the (now changed) `UserInfo` class. I was confused about using both User classes if we imported them to the same place since we could get errors or stuff would be overwritten. They do have different constructors, so maybe it would have been accepted? Not sure, either way changed for the sake of readability and to save potential confusion. The set up makes more sense now, thanks

Fixed